### ISSUE TITLE
A paused simulation gets unpaused after changing node position

### ISSUE THREAD

**Issue by Tisten** [OurMachinery/themachinery-public#1237]
- **Version:** Git Master Branch
- **OS:** Windows

Repro project: 
[PausedSimulationBug.zip](https://github.com/OurMachinery/themachinery-public/files/8391166/PausedSimulationBug.zip)

Steps:
1. Make sure your speakers/headphones are not muted.
2. After opening the project, start a simuation of the world, this will start looping a sound
3. Open the entity graph in the world entity. This will pause the simulation and stop playing the sound.
4. Move a node a few pixels. Now the sound plays again, even though the simulation tab says paused. The init event got triggered, and nothing paused it after.
5. Move a node again. Now the sound is playing twice on top of each other.

**Comment by niklas-ourmachinery** [comment]
Ok, this one is a bit tricky... there's a bunch of things going on here and they are all in the category where you can debate whether they are bugs or intended behaviors, so let's go through them one-by-one.

**Init triggers even though the simulation is paused**

Currently, what Pause means is that Tick events don't run, Init events still run. I *think* this is the behavior we want.

**Init triggers every time a graph is reloaded**

When an asset used in a simulation is changed, we don't restart the whole simulation, we just reload that particular asset. I think that is what you want, because most of the time you want to keep the state of the game and just see the small change that you did. And if you want a full restart you can always use the restart button.

When a graph component is reloaded, *Init* is triggered -- this is debatable. On the one hand, triggering Init means that any setup that the graph is doing happens multiple times. On the other hand, not triggering Init means that changes to the graph that rely on something happening in Init won't be seen in the simulation until the reload.

I think not triggering Init on reload is the right thing to do... it seems like having the initialization run twice is a worse problem than not having some initialization code run... but I'm not sure.

@leonardo-ourmachinery I seem to remember we had a discussion about this a while ago, but I don't remember what our conclusion was.

**Sound plays even though the simulation is paused**

The simulation API causes `tm_mixer_api->pause_world()` to pause all currently playing sounds in the world when the pause button is played. However, you can still play *new sounds* and they won't be paused. I think this should be changed so that if you play new sounds while the world is paused, they will start in a paused state and not play.

**Comment by Tisten** [comment]
> if you play new sounds while the world is paused, they will start in a paused state and not play.

I too think that this is an important part of the fix.

Secondly, I would expect a "shut down" of the graph, i.e anything spawned on "init" should be freed before the next init, no matter which path the code takes from one init to the next. Otherwise any manipulation of the graph can create a form of resource leak.

In general I would expect all systems to respect the "paused" state, no matter if they are particle systems, physics, animations, sound or gamepad rumble. And I guess I would even expect them to respect slow motion as well, i.e that the simulation speed should be driven by a delta time. It can become a problem to pass in zero if anything is divided by time, but that case can be a great integration test to use.

So in summary, I'm fine with the graph getting multiple init calls, but it shouldn't start "simulating", and it should be shut down before the next init.

**Comment by leonardo-ourmachinery** [comment]
@niklas-ourmachinery yes, we talked about that one a while ago when I was dealing with reloading assets from the gamestate, we decided to postpone the decision and see what happens when users start to play around with this init-reload thing a bit more... and here we go.

I think the correct decision is to have the init event only run when the asset is created.
If the users want to redo the things that are done in the init event when the asset is reloaded, he can just link the reload event to it.

---

@Tisten current we're doing this, which I think is the correct thing to do:
```
double dt = sim->paused ? 0.0f : wall_dt * simulation_speed;
[...]
tm_entity_api->set_blackboard_double(sim->entity_ctx, TM_ENTITY_BB__DELTA_TIME, dt);
[...]
tm_entity_api->propagate_asset_changes(sim->entity_ctx);
if (!sim->paused)
        tm_entity_api->update(sim->entity_ctx);
```
When you change the position of the node in the graph, `propagate_asset_changes()` is detecting that even if the simulation is paused, and the components of the modified entity are reloaded.
We could argue that we shouldn't propagate asset changes if the simulation is paused too... @niklas-ourmachinery what do you think?
I'll think about that a bit more, in the meanwhile you can try to see how it feels locally if you want to.

---

My feeling is that if we add those two things then there's no need to check if the simulation is paused or not when we play a sound, quite the contrary we could _assume_ that the simulation is not in a paused state. (If you are executing a graph node it means you are inside either the `propagate_asset_changes()` or the `update()` calls, but those will only be executed if the simulation is not paused.

**Comment by leonardo-ourmachinery** [comment]
I've pushed the fixes here: https://github.com/OurMachinery/themachinery/commit/7b8213f766401abc54d07c45d1eac6125e813529

Both with your project and with our physics sample project now everything feels more intuitive to me, I think it's the right decision.
There's only one use case which I'd like to discuss, and that is that now if you spawn something in the reload event, you won't see it until the simulation is unpaused... I think this is correct, the fact is that you won't _see_ it until you explicitly press the unpause button. To me seems the correct behavior, but I'm not 100% sure.

Also worth noting: if you want to execute again the init phase of the entity you can now:
-pause the simulation
-create the reload event in the graph and link it to what you wanna do
-unpause the simulation
-Though now you have to remember to delete the reload node, as otherwise it will keep doing that every time.

**Comment by niklas-ourmachinery** [comment]
I agree... let's have `Init` just run on creation and use `Reload` for reload initializations.

I'm a bit on the fence about propagating changes in the "paused" state. To me, "paused" means that just the gameplay is paused and you would still see edits done to the level. But it might be simpler in many ways to just wait with processing changes until we're unpaused.

I think it could be argued either way... maybe just try this out and see how it feels.

**Comment by leonardo-ourmachinery** [comment]
Yeah, I tested this a bit and I think this is the way to go: don't propagate changes if the simulation is paused.
This supports the following workflow:
-Pause the simulation
-Open the graph of a specific entity
-Script the Reload event by adding as many nodes as you want
-Unpause the simulation: The Reload event will only run once.

If we don't do this, The reload event will be run every time you add a new node to it, which I don't think is what users expect if the simulation is paused.

Also, I think this makes the code a bit more robust too as we can now ensure that the entity context won't be touched at all if the simulation is paused.

Closing this for now, @Tisten feel free to reopen if you disagree or if there's anything else you want to discuss regarding this.
